### PR TITLE
[runtime][source info][mac] return "inline at" position for stack trace

### PR DESCRIPTION
**Pull request by vvlevchenko**

### CODE REVIEW

File: runtime/src/debug/cpp/SourceInfo.cpp [hunk 1]
```diff
@@ -22,6 +22,17 @@
 #include <limits.h>
 #include <stdint.h>
 #include <unistd.h>
+#include <string.h>
+
+#define TRACE_SYMBOLICATION 1
```
Comment (SvyatoslavScherbina) [inline runtime/src/debug/cpp/SourceInfo.cpp:27]: Should probably be `0`.

File: runtime/src/debug/cpp/SourceInfo.cpp [hunk 2]
```diff
@@ -22,6 +22,17 @@
...
+#if TRACE_SYMBOLICATION
+#include <stdio.h>
+#define SYM_LOG(...) fprintf(stderr, __VA_ARGS__)
+#define SYM_DUMP(p) CSShow((p))
+#else
+#define SYM_LOG(...)
+#define SYM_DUMP(p) CSShow((p))
```
Comment (SvyatoslavScherbina) [inline runtime/src/debug/cpp/SourceInfo.cpp:34]: Should be empty.

File: runtime/src/debug/cpp/SourceInfo.cpp [hunk 3]
```diff
@@ -114,22 +143,61 @@ extern "C" struct SourceInfo Kotlin_getSourceInfo(void* addr) {
...
+     * ASSUMPTION: we assume that the _first_ and the _last_ source infos should belong to real function(symbol) the rest might belong to
+     * inlined functions.
+     */
+    CSSymbolForeachSourceInfo(symbol,
+          ^(CSSourceInfoRef ref) {
+           uint32_t lineNumber = CSSourceInfoGetLineNumber(ref);
+           if (lineNumber == 0)
+             return 0;
+           if (limits.start == -1) {
+              limits.start = lineNumber;
```
Comment (SvyatoslavScherbina) [inline runtime/src/debug/cpp/SourceInfo.cpp:159]: Indent has 3 spaces.

File: runtime/src/debug/cpp/SourceInfo.cpp [hunk 4]
```diff
@@ -114,22 +143,61 @@ extern "C" struct SourceInfo Kotlin_getSourceInfo(void* addr) {
...
+              limits.fileName = CSSourceInfoGetPath(ref);
+           } else
+              limits.end = lineNumber;
```
Comment (SvyatoslavScherbina) [inline runtime/src/debug/cpp/SourceInfo.cpp:162]: No `{}` for `else` branch, while `if` has `{}`.

File: backend.native/tests/runtime/exceptions/kt-37572.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        null,
+        "kt-37572.kt" to 25,
+        "kt-37572.kt" to 16,
+        "kt-37572.kt" to 6,
+        "kt-37572.kt" to 4)
+
+internal fun checkFrame(value:String) {
+    val (pos, file, line) = regex.find(value)!!.destructured
+    goldValues[pos.toInt()]?.let{
+        assertTrue(file == it.first && line.toInt() == it.second, "expected ${it.first}:${it.second} got $file:$line")
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/exceptions/kt-37572.kt:39]: Why not just have two `assertEquals` instead?
`assertEquals` has proper message by default.

File: backend.native/tests/runtime/exceptions/kt-37572.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        null,
+        "kt-37572.kt" to 25,
+        "kt-37572.kt" to 16,
+        "kt-37572.kt" to 6,
+        "kt-37572.kt" to 4)
+
+internal fun checkFrame(value:String) {
+    val (pos, file, line) = regex.find(value)!!.destructured
+    goldValues[pos.toInt()]?.let{
+        assertEquals(file, it.first)
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/exceptions/kt-37572.kt:39]: Arguments should probably be flipped: expected value must be the first.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 7]
```diff
@@ -1703,17 +1700,17 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
 
         override fun returnableBlockScope(): CodeContext? = this
 
-        override fun location(line: Int, column: Int): LocationInfo? {
+        override fun location(offset: Int): LocationInfo? {
             return if (returnableBlock.inlineFunctionSymbol != null) {
                 val diScope = functionScope ?: return null
                 val outerFileEntry = outerFileEntry()
-                val inlinedAt = outerContext.location(outerFileEntry.line(returnableBlock.startOffset), outerFileEntry.column(returnableBlock.startOffset))
+                val inlinedAt = outerContext.location(outerFileEntry.line(returnableBlock.startOffset))
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1707]: Is this correct?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 8]
```diff
@@ -684,7 +680,8 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
             declaration?.scope() ?: llvmFunction!!.scope(0, subroutineType(context, codegen.llvmTargetData, listOf(context.irBuiltIns.intType)))
         }
 
-        override fun location(line: Int, column: Int) = scope?.let { LocationInfo(it, line, column) }
+        private val fileScope = (fileScope() as? FileScope)
+        override fun location(offset: Int) = scope?.let { scope -> fileScope?.let{LocationInfo(scope, it.file.fileEntry.line(offset), it.file.fileEntry.column(offset)) } }
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:684]: What exactly does this fix?

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:684]: I mean that probably we've got wrong file scope when calling `line`/`column` at call site. Could you point me at the specific code fragment that had wrong file scope?

Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:684]: no we have it right for most of the block (it's returnable block), but when we calculate location for variable (captured by inliner) we could be in the middle of returnable block and file isn't actual for variable belonging to call site.    this exactly the place we used wrong place to calculate line/column position

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 7 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1707]: yep, we calculate real line/column position using actual file

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1707]: `location` expects `offset` but gets line number.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 8 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:684]: That's what I meant :)
Is it code context from `switchSymbolizationContextTo` that was affected by the fix?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 7 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1707]: Oh, that definitely shouldn't be in this way

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 8 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:684]: yes

**Pull request merged**